lightning/src/ln/features.rs

```diff
@@ -0,0 +1,251 @@
+	impl FeatureContext for FeatureContextChannel {}
+	//impl FeatureContext for FeatureContextInvoice {}
+
+	pub trait DataLossProtect: FeatureContext {}
+	impl DataLossProtect for FeatureContextInit {}
+	impl DataLossProtect for FeatureContextNode {}
+
+	pub trait InitialRoutingSync: FeatureContext {}
+	impl InitialRoutingSync for FeatureContextInit {}
+
+	pub trait UpfrontShutdownScript: FeatureContext {}
+	impl UpfrontShutdownScript for FeatureContextInit {}
+	impl UpfrontShutdownScript for FeatureContextNode {}
+}
+
+/// Tracks the set of features which a node implements, templated by the context in which it
+/// appears.
+pub struct Features<T: sealed::FeatureContext> {
+	#[cfg(not(test))]
+	/// Note that, for convinience, flags is LITTLE endian (despite being big-endian on the wire)
+	flags: Vec<u8>,
+	// Used to test encoding of diverse msgs
+	#[cfg(test)]
+	pub flags: Vec<u8>,
+
+	#[cfg(not(test))]
+	mark: PhantomData<T>,
+	#[cfg(test)]
+	pub mark: PhantomData<T>,
+}
+
+impl<T: sealed::FeatureContext> Clone for Features<T> {
+	fn clone(&self) -> Self {
+		Self {
+			flags: self.flags.clone(),
+			mark: PhantomData,
+		}
+	}
+}
+impl<T: sealed::FeatureContext> PartialEq for Features<T> {
+	fn eq(&self, o: &Self) -> bool {
+		self.flags.eq(&o.flags)
+	}
+}
+impl<T: sealed::FeatureContext> fmt::Debug for Features<T> {
+	fn fmt(&self, fmt: &mut fmt::Formatter) -> Result<(), fmt::Error> {
+		self.flags.fmt(fmt)
+	}
+}
+
+/// A feature message as it appears in an init message
+pub type InitFeatures = Features<sealed::FeatureContextInit>;
+/// A feature message as it appears in a node_announcement message
+pub type NodeFeatures = Features<sealed::FeatureContextNode>;
+/// A feature message as it appears in a channel_announcement message
+pub type ChannelFeatures = Features<sealed::FeatureContextChannel>;
+
+impl InitFeatures {
+	/// Create a Features with the features we support
+	#[cfg(not(feature = "fuzztarget"))]
+	pub(crate) fn our_features() -> InitFeatures {
+		InitFeatures {
+			flags: vec![2 | 1 << 5],
+			mark: PhantomData,
+		}
+	}
+	#[cfg(feature = "fuzztarget")]
+	pub fn our_features() -> InitFeatures {
+		InitFeatures {
+			flags: vec![2 | 1 << 5],
+			mark: PhantomData,
+		}
+	}
+
+	/// Writes all features present up to, and including, 13.
+	pub(crate) fn write_up_to_13<W: Writer>(&self, w: &mut W) -> Result<(), ::std::io::Error> {
+		let len = cmp::min(2, self.flags.len());
+		w.size_hint(len + 2);
+		(len as u16).write(w)?;
+		for i in (0..len).rev() {
+			if i == 0 {
+				self.flags[i].write(w)?;
+			} else {
+				(self.flags[i] & ((1 << (14 - 8)) - 1)).write(w)?;
+			}
+		}
+		Ok(())
+	}
+
+	/// or's another InitFeatures into this one.
+	pub(crate) fn or(&mut self, o: &InitFeatures) {
+		let total_feature_len = cmp::max(self.flags.len(), o.flags.len());
+		self.flags.resize(total_feature_len, 0u8);
+		for (feature, o_feature) in self.flags.iter_mut().zip(o.flags.iter()) {
+			*feature |= *o_feature;
+		}
+	}
+}
```

Comment:
Hmmmm...at least personal-preference-wise I hate overloading operators. I guess you could argue that Features intends to be a bitfield so you could argue it really *is* an arithmetic type and it makes sense, but man I hate it.

Reply:
Somewhat agree with you on operator overloading. I suppose the underlying reason behind why I suggested it is that I think its interface is more suitable (i.e., because it doesn't mutate `self`). Then the call site would be pretty clear that you are merging the global and local features. If using `mut self`, then naming the method as `merge` would be clearer.

Reply:
Yea, I spent a bit thinking about a better function name and couldn't come up with one. One alternative is to move the self from a reference to a non-reference. We could even take both sides by non-reference and return one of them, which would optimize pretty similarly but prevent any confused usage.

Reply:
SGTM